### PR TITLE
Finish AST types

### LINKED ISSUE #25: Design and implement AST data structure

### PR DESCRIPTION
**Pull request by rwblickhan**
:tickets: **Ticket(s)**: Closes #25

---

## :construction_worker: Changes

Create all AST types. Also hacked around with @bobheadxi's work to make it all compile... we'll have to clean that up in the future.
 
## :flashlight: Testing Instructions

It compiles, I guess?

### CODE REVIEW

**Review by bobheadxi — APPROVED** [review]
not sure how I feel about all the shortenings (`Expr` vs `Expression` - I think verbosity is fine, especially when IDE's can autocomplete - but that's a purely cosmetic complaint

**Comment by bobheadxi** [comment]
Namespace?

File: src/include/ASTree.h [hunk 1]
```diff
@@ -6,106 +6,196 @@
...
+
+struct UnaryExpr : public Expr {
+    TokenType m_op;
+    std::unique_ptr<UnaryExpr> m_operand;
+    UnaryExpr(TokenType op, std::unique_ptr<UnaryExpr> operand)
+        : m_op(op)
+        , m_operand(std::move(operand)) { }
+};
+
+struct NoOpExpr : public Expr { };
```
Comment (CoffeeTableEspresso) [inline src/include/ASTree.h]: Remove this.

File: src/include/ASTree.h [hunk 2]
```diff
@@ -6,106 +6,196 @@
...
 
-struct Literal : public Expression { };
+struct Literal : public Expr { };
```
Comment (CoffeeTableEspresso) [inline src/include/ASTree.h]: Remove this too.

File: src/include/ASTree.h [hunk 3]
```diff
@@ -6,106 +6,196 @@
...
+    MulExpr(TokenType op, std::unique_ptr<UnaryExpr> left, std::unique_ptr<UnaryExpr> right)
+        : m_op(op)
+        , m_left(std::move(left))
+        , m_right(std::move(right)) { }
+};
+
+struct UnaryExpr : public Expr {
+    TokenType m_op;
+    std::unique_ptr<UnaryExpr> m_operand;
+    UnaryExpr(TokenType op, std::unique_ptr<UnaryExpr> operand)
```
Comment (CoffeeTableEspresso) [inline src/include/ASTree.h]: Make `explicit`, see others.

File: src/include/ASTree.h [hunk 4]
```diff
@@ -6,106 +6,196 @@
...
-struct UnaryExpression : public Expression {
+struct CompExpr : public Expr {
     TokenType m_op;
-    std::unique_ptr<Expression> m_operand;
-public:
-    UnaryExpression(TokenType op, std::unique_ptr<Expression> operand)
-            : m_op(op)
-            , m_operand(std::move(operand)) { }
+    std::unique_ptr<ConcatExpr> m_left;
+    std::unique_ptr<ConcatExpr> m_right;
```
Comment (CoffeeTableEspresso) [inline src/include/ASTree.h]: Add explicit constructor, see others

File: src/include/ASTree.h [hunk 5]
```diff
@@ -6,106 +6,196 @@
...
+        , m_operand(std::move(operand)) { }
+};
+
+struct NoOpExpr : public Expr { };
 
-struct Literal : public Expression { };
+struct Literal : public Expr { };
 
 struct Integer : public Literal {
     int m_value;
```
Comment (CoffeeTableEspresso) [inline src/include/ASTree.h]: use `int64_t`

File: src/include/ASTree.h [hunk 6]
```diff
@@ -6,106 +6,196 @@
...
-public:
     explicit Integer(const std::string &value) : m_value(std::stoi(value)) { }
```
Comment (CoffeeTableEspresso) [inline src/include/ASTree.h]: use `stoll`.

File: src/include/ASTree.h [hunk 7]
```diff
@@ -6,106 +6,196 @@
...
 };
 
 struct Float : public Literal {
     float m_value;
-public:
     explicit Float(const std::string &value) : m_value(std::stof(value)) { }
```
Comment (CoffeeTableEspresso) [inline src/include/ASTree.h]: use `stod`

File: src/include/ASTree.h [hunk 8]
```diff
@@ -6,106 +6,196 @@
...
+struct Literal : public Expr { };
 
 struct Integer : public Literal {
     int m_value;
-public:
     explicit Integer(const std::string &value) : m_value(std::stoi(value)) { }
 };
 
 struct Float : public Literal {
     float m_value;
```
Comment (CoffeeTableEspresso) [inline src/include/ASTree.h]: use `double`

File: src/include/ASTree.h [hunk 9]
```diff
@@ -6,106 +6,196 @@
...
+};
+
 //////////////////////////////////////////
 /// Expressions
 
-struct Expression : public Statement { };
+struct Expr : public Statement { };
+
+struct OrExpr : public Expr {
+    std::unique_ptr<AndExpr> m_left;
```
Comment (CoffeeTableEspresso) [inline src/include/ASTree.h]: make these children `Expr`, see others.

**Review by CoffeeTableEspresso — CHANGES_REQUESTED** [review]

File: src/include/ASTree.h (hunk 1 above)
Comment (rwblickhan) [reply src/include/ASTree.h]: Fixed.

File: src/include/ASTree.h (hunk 2 above)
Comment (rwblickhan) [reply src/include/ASTree.h]: Fixed.

File: src/include/ASTree.h (hunk 3 above)
Comment (rwblickhan) [reply src/include/ASTree.h]: Fixed.

File: src/include/ASTree.h (hunk 4 above)
Comment (rwblickhan) [reply src/include/ASTree.h]: Fixed.

File: src/include/ASTree.h (hunk 9 above)
Comment (rwblickhan) [reply src/include/ASTree.h]: Fixed.

File: src/include/ASTree.h (hunk 8 above)
Comment (rwblickhan) [reply src/include/ASTree.h]: Fixed.

File: src/include/ASTree.h (hunk 7 above)
Comment (rwblickhan) [reply src/include/ASTree.h]: Fixed.

File: src/include/ASTree.h (hunk 6 above)
Comment (rwblickhan) [reply src/include/ASTree.h]: Fixed.

File: src/include/ASTree.h (hunk 5 above)
Comment (rwblickhan) [reply src/include/ASTree.h]: Fixed.

File: src/include/ASTree.h [hunk 10]
```diff
@@ -6,106 +6,239 @@
...
-};
-
-struct Return : public Statement {
-    std::unique_ptr<Expression> m_expression;
-public:
-    explicit Return(std::unique_ptr<Expression> expression)
-        : m_expression(std::move(expression)) { }
+struct Group : public Expr {
+    std::unique_ptr<Expr> m_expr;
+    explicit Group(std::unique_ptr<Expr> expr) : m_expr(std::move(expr)) { }
```
Comment (CoffeeTableEspresso) [inline src/include/ASTree.h]: delete Group.

**Review by CoffeeTableEspresso — DISMISSED** [review]
Looks good, just need 1 change.

File: src/include/ASTree.h (hunk 10 above)
Comment (rwblickhan) [reply src/include/ASTree.h]: Fixed.

**Review by bobheadxi — APPROVED** [review]

File: src/parser/Parser.cpp [hunk 11]
```diff
@@ -94,39 +84,39 @@ std::unique_ptr<Expression> Parser::parseExpression() {
...
-                    String(m_currentToken.literal));
+            statement = util::make_unique<AST::StringLit>(
+                    AST::StringLit(m_currentToken.literal));
             break;
         case TokenType::True:
         case TokenType::False:
-            statement = util::make_unique<Boolean>(
-                    Boolean(m_currentToken.literal));
+            statement = util::make_unique<AST::BooleanLit>(
+                    AST::BooleanLit(m_currentToken.literal));
```
Comment (bobheadxi) [inline src/parser/Parser.cpp:105]: we lit now boiz

Comment (rwblickhan) [reply src/parser/Parser.cpp:105]: 🔥

**Pull request merged**